query_condition_test.go

```diff
@@ -0,0 +1,414 @@
+	array, err := createBasicTestArray("test_query_condition")
+	if err != nil {
+		t.Errorf("failed to create basic test array: %s", err)
+	}
+	defer os.RemoveAll(array.uri)
+
+	if err := array.Open(TILEDB_READ); err != nil {
+		t.Errorf("failed to open test array for reading: %s", err)
+	}
+
+	testQueryConditionInt32(t, array)
+	testQueryConditionBytes(t, array)
+	testQueryConditionTime(t, array)
+}
+
+func testQueryConditionInt32(t *testing.T, array *Array) {
+	a1Cases := []struct {
+		name           string
+		opValue        int32
+		op             QueryConditionOp
+		expectedValues []int32
+	}{
+		{"GreaterThan1", 1, TILEDB_QUERY_CONDITION_GT, []int32{2, 3}},
+		{"LessThan3", 3, TILEDB_QUERY_CONDITION_LT, []int32{1, 2}},
+		{"EqualTo2", 2, TILEDB_QUERY_CONDITION_EQ, []int32{2}},
+	}
+	for _, c := range a1Cases {
+		c := c
+		t.Run(c.name, func(t *testing.T) {
+			a1DataRead := make([]int32, 3)
+			// Prepare the query
+			query, err := NewQuery(array.context, array)
+			assert.Nil(t, err)
+			assert.NotNil(t, query)
+
+			_, err = query.SetBuffer("a1", a1DataRead)
+			assert.Nil(t, err)
+			assert.NotNil(t, query)
+
+			qc, err := NewQueryCondition(array.context, "a1", c.op, c.opValue)
+			assert.Nil(t, err)
+
+			err = query.SetQueryCondition(qc)
+			assert.Nil(t, err)
+
+			err = query.SetLayout(TILEDB_ROW_MAJOR)
+			assert.Nil(t, err)
+
+			// Submit the query
+			err = query.Submit()
+			assert.Nil(t, err)
+			// compare the elements in the buffer to the expected values
+			elements, err := query.ResultBufferElements()
+			assert.Nil(t, err)
+
+			for i := uint64(0); i < elements["a1"][1]; i++ {
+				assert.Equal(t, a1DataRead[i], c.expectedValues[i])
+			}
+			query.Free()
+		})
+	}
+
+	a1CombinationCases := []struct {
+		name                 string
+		op1                  QueryConditionOp
+		op1Value             int32
+		op2                  QueryConditionOp
+		op2Value             int32
+		combinationCondition QueryConditionCombinationOp
+		expectedValues       []int32
+	}{
+		{
+			name:                 "GreaterThan1AndLessThan3",
+			op1:                  TILEDB_QUERY_CONDITION_GT,
+			op1Value:             1,
+			op2:                  TILEDB_QUERY_CONDITION_LT,
+			op2Value:             3,
+			combinationCondition: TILEDB_QUERY_CONDITION_AND,
+			expectedValues:       []int32{2},
+		},
+	}
+	for _, c := range a1CombinationCases {
+		c := c
+		t.Run(c.name, func(t *testing.T) {
+			a1DataRead := make([]int32, 3)
+			// Prepare the query
+			query, err := NewQuery(array.context, array)
+			assert.Nil(t, err)
+			assert.NotNil(t, query)
+
+			_, err = query.SetBuffer("a1", a1DataRead)
+			assert.Nil(t, err)
+			assert.NotNil(t, query)
+
+			qc1, err := NewQueryCondition(array.context, "a1", c.op1, c.op1Value)
+			assert.Nil(t, err)
+
+			qc2, err := NewQueryCondition(array.context, "a1", c.op2, c.op2Value)
+			assert.Nil(t, err)
+
+			qc, err := NewQueryConditionCombination(array.context, qc1, c.combinationCondition, qc2)
```

Comment:
Since `NewQueryConditionCombination` does not alter existing conditions, isn't this line's effect shadowed by above two lines (`NewQueryCondition`)?

Reply:
The NewQueryConditionCombination functionality just merges the two conditions together and doesn't mutate the underlying query conditions. In effect it creates a new query condition with both predicates in one "query condition". We then use only this query condition when doing the actual query

Reply:
👍